cli/cliui/select.go

```diff
@@ -75,33 +75,172 @@ func Select(inv *serpent.Invocation, opts SelectOptions) (string, error) {
 		return opts.Options[0], nil
 	}
 
-	var defaultOption interface{}
-	if opts.Default != "" {
-		defaultOption = opts.Default
+	initialModel := selectModel{
+		search:     textinput.New(),
+		hideSearch: opts.HideSearch,
+		options:    opts.Options,
+		height:     opts.Size,
+		message:    opts.Message,
 	}
 
+	if initialModel.height == 0 {
+		initialModel.height = 7
+	}
+
+	initialModel.search.Prompt = ""
+	initialModel.search.Focus()
+
+	m, err := tea.NewProgram(
+		initialModel,
+		tea.WithInput(inv.Stdin),
+		tea.WithOutput(inv.Stdout),
+	).Run()
+
 	var value string
-	err := survey.AskOne(&survey.Select{
-		Options:  opts.Options,
-		Default:  defaultOption,
-		PageSize: opts.Size,
-		Message:  opts.Message,
-	}, &value, survey.WithIcons(func(is *survey.IconSet) {
-		is.Help.Text = "Type to search"
-		if opts.HideSearch {
-			is.Help.Text = ""
-		}
-	}), survey.WithStdio(fileReadWriter{
-		Reader: inv.Stdin,
-	}, fileReadWriter{
-		Writer: inv.Stdout,
-	}, inv.Stdout))
-	if errors.Is(err, terminal.InterruptErr) {
-		return value, Canceled
+	if m, ok := m.(selectModel); ok {
+		if m.canceled {
+			return value, Canceled
+		}
+		value = m.selected
 	}
 	return value, err
 }
 
+type selectModel struct {
+	search     textinput.Model
+	options    []string
+	cursor     int
+	height     int
+	message    string
+	selected   string
+	canceled   bool
+	hideSearch bool
+}
+
+func (selectModel) Init() tea.Cmd {
+	return textinput.Blink
+}
+
+//nolint:revive // The linter complains about modifying 'm' but this is typical practice for bubbletea
+func (m selectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var cmd tea.Cmd
+
+	if msg, ok := msg.(tea.KeyMsg); ok {
```

Comment:
What happens if the program receives a `kill -INT $PID` while tea is active? Do we handle it nicely?

Reply:
Unless I'm doing something wrong, it appears nothing happens?

Reply:
Hmm, did you try with a `coder` command or cliui test binary? With `coder` I would expect the program to exit with a cancelled.

Reply:
I used the `cliui` binary with `go run ./cmd/cliui select`. I shall try against the coder binary

Reply:
Ah so running it without `go run` changes the behaviour. If the caller doesn't handle the `Canceled` error then it returns an empty string for the `Select` component

Reply:
> Ah so running it without go run changes the behaviour.
Yeah go run compiles the binary and it runs with a different PID than the actual `go run` command.
> If the caller doesn't handle the Canceled error then it returns an empty string for the Select component
This is what I would expect 👍🏻

Reply:
Have just checked on the `MultiSelect` component, it returns the currently selected options.

Reply:
Does it populate error as well (it should)?

Reply:
Just tested and it does not! Need to handle that then, `err` is `nil`

Reply:
I can confirm by disabling the builtin signal handler it is possible to get `err` to contain the `Canceled` error

Reply:
Might I suggest returning a func instead of the channel for close/deregistering the handler? I feel the usage will be a bit more idiomatic/easy to understand.

On the other hand, perhaps we don't need to handle signals at all and leave it to be the responsibility of the `coder` command that needs it? I guess it's possible the terminal is left in a bad state unless the signal is handled though (not sure if it enters raw mode or not).

PS. You might want to only handle `signal.Interrupt` for the Windows compatibility.

Reply:
Yes, definitely a good idea. Have made that change 👍

Killing the process with `kill -TERM $PID` leaves the cursor in a bad state if we don't explicitly handle it (`syscall.SIGTERM`).

Will push these changes up, can always revert if unhappy.